### DISCUSSION TITLE
Box selection from catalog in settings

### DISCUSSION THREAD

**Discussion by RitsmG** [OFFLINE-GmbH/oc-boxes-support#95] [Ideas and feature requests]
Hello.

We have a large selection of boxes created for our theme which we re-use for all our clients. But we don't want all boxes to be "enabled" for each site because we don't always style them all for the new site or they are generally not needed for the type of site we create. Al the boxes do create a rather large list of boxes we can add to the page. 

Could it be possible if there was some sort of filter/selection function, in the settings for boxes, in which we can select the boxes we want available for the current site we are working on?

**Comment by tobias-kuendig** [comment]
Where are the boxes stored? In a global theme? This would mean that you cannot change the Boxes themselfes but you need to have full control on what to show for each October installation?

I'm asking because we already have [contexts](https://docs.boxes.offline.ch/reference/schema.html#contexts). We could add a setting to override the current `default` content that the backend editor uses and allow you to define what contexts to use. This way, you could enable only certain kinds of Boxes. 

But your use-case probably requires you to enable/disable each individual Box?

How about an event that allows you to access the full Boxes list (to generate a backend form/checkbox list) and another event that allows you to filter the Boxes before they are displayed in the Editor? This way you could add the required management features in a fairly simple custom plugin.

**Reply by RitsmG** [reply]
_"How about an event that allows you to access the full Boxes list (to generate a backend form/checkbox list) and another event that allows you to filter the Boxes before they are displayed in the Editor? This way you could add the required management features in a fairly simple custom plugin."_

This would actually be a really good fix for our use-case.

This would enable us to select from all boxes individually or create sets of boxes for specific kinds of sites.

EDIT:
For some extra information. Our boxes are indeed stored in a global theme which we overwrite by using the child theme function.

**Comment by tobias-kuendig** [comment]
Version 3.3.14 comes with two new events: `offline.boxes.beforeFilterPartials` and `offline.boxes.filterPartials`. Let me know if this fits your use-case.

Docs: https://docs.boxes.offline.ch/reference/events.html

**Comment by RitsmG** [comment]
Hello @tobias-kuendig ,

When using the `offline.boxes.filterPartials` the filtered out boxes are indeed not available for use. Though they do appear in the boxes selection popup when adding a new box to the page. As we read the documentation this should not be the case.

**Reply by gldrenthe89** [reply]
@tobias-kuendig i also tried this. The boxes selections popup shows al boxes, even though i almost 'forgot' every box available. When selection a box that is not available but visible an error is thrown. Looks like the event is not triggered when openening the boxes popup window.

**Reply by tobias-kuendig** [reply]
I was able to reproduce this issue, thank you for reporting it! The backend currently uses a separate method to list the available partials that circumvent the event.

I'll release a fix for this tomorrow.

**Reply by gldrenthe89** [reply]
@tobias-kuendig Thanks for a very quick reply 🙏

**Reply by tobias-kuendig** [reply]
Please give v3.3.20 a try, this should be fixed now. Thanks!

**Reply by gldrenthe89** [reply]
Hi @tobias-kuendig sorry for the late response. We are waiting on internal development to be able to update to this version. This will be done this afternoon. Keep you posted.

**Comment by gldrenthe89** [comment]
Hi @tobias-kuendig v3.3.20 did it. it is fixed now. Many thanks for this quick update again. 

Just a follow-up quastion. Is there a way for us the use the PartialReader class without triggering the Events. We wanna use this function in a Setting Model where we per customer select the boxes that are available for that specific customer. 

Now i have fixed this with a class that extends PartialReader and copied the init() function and removed the event triggers. Short, we just want the full list with all 'handles' that are in the code.

We have it working now. But maybe this can be done way easier.

**Reply by tobias-kuendig** [reply]
@gldrenthe89 the beforeFilter event contains all partials: https://github.com/OFFLINE-GmbH/oc-boxes-free-plugin/blob/main/classes/partial/PartialReader.php#L148

You would have to listen for it and store the result somewhere. Maybe this can be a solution?

**Reply by gldrenthe89** [reply]
That could be a workaround. Thanks
